src/devices/bus/coco/coco_ram.cpp

```diff
@@ -0,0 +1,183 @@
+
+	protected:
+		// device-level overrides
+		virtual void device_start() override;
+		virtual void device_reset() override;
+		virtual DECLARE_WRITE8_MEMBER(scs_write) override;
+		virtual DECLARE_READ8_MEMBER(scs_read) override;
+
+	private:
+		required_device<ram_device>             m_staticram;
+		int										m_offset;
+	};
+};
+
+
+
+//**************************************************************************
+//  GLOBAL VARIABLES
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE_PRIVATE(COCO_PAK_RAM, device_cococart_interface, coco_pak_ram_device, "cocopakram", "Disto " STRINGIZE(RAM_SIZE_IN_K) "K RAM Cartridge")
+
+
+
+//-------------------------------------------------
+//  coco_pak_device - constructor
+//-------------------------------------------------
+
+coco_pak_ram_device::coco_pak_ram_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, COCO_PAK_RAM, tag, owner, clock)
+	, device_cococart_interface(mconfig, *this)
+	, m_staticram(*this, STATICRAM_TAG)
+	, m_offset(0)
+{
+}
+
+
+
+//**************************************************************************
+//  MACHINE FRAGMENTS AND ADDRESS MAPS
+//**************************************************************************
+
+void coco_pak_ram_device::device_add_mconfig(machine_config &config)
+{
+	RAM(config, STATICRAM_TAG).set_default_size(STRINGIZE(RAM_SIZE_IN_K) "K").set_default_value(0);
+}
+
+
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void coco_pak_ram_device::device_start()
+{
+	// initial state
+	m_offset = 0;
+
+	// save state
+	save_item(NAME(m_offset));
+}
+
+
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void coco_pak_ram_device::device_reset()
+{
+	m_offset = 0;
+}
+
+
+
+//-------------------------------------------------
+//    scs_write
+//-------------------------------------------------
+
+WRITE8_MEMBER(coco_pak_ram_device::scs_write)
+{
+// 	int idata = data;
+
+	switch(offset)
+	{
+		case 0:
+			m_offset = ((m_offset & 0xffff00) + data);
+			break;
+		case 1:
+			m_offset = ((m_offset & 0xff00ff) + (data << 8));
+			break;
+		case 2:
+			m_offset = ((m_offset & 0x00ffff) + (data << 16));
+			break;
+		case 3:
+			if( m_offset < BUFFER_SIZE )
+			{
+				m_staticram->write(m_offset, data);
+			}
+			break;
+	}
+
+	LOG("scs_write: %s: %06x, %02x, %02x\n", machine().describe_context(), m_offset, offset, data);
+}
+
+
+
+//-------------------------------------------------
+//  scs_read
+//-------------------------------------------------
+
+READ8_MEMBER(coco_pak_ram_device::scs_read)
+{
+	uint8_t data = 0x00;
+
+	switch (offset)
+	{
+		case 0:
+			data = (m_offset) & 0xff;
+			break;
+		case 1:
+			data = (m_offset & 0xff00ff) >> 8;
```

Comment:
This can't possibly work – you’re shifting the part you masked out into the result.

Reply:
mask is inverted, easy fix

Reply:
I've retested with the official testing program and it still works.